x/mongo/driver/topology/topology.go

```diff
@@ -605,7 +617,11 @@ func (t *Topology) apply(ctx context.Context, desc description.Server) descripti
 	var current description.Topology
 	var err error
 	current, desc, err = t.fsm.apply(desc)
-	if err != nil {
+
+	if err != nil && current.CompatibilityErr != nil {
```

Comment:
This seems risky. We're now using `t.Description()` rather than the actual topology description being stored in the FSM. We're also changing the behavior for a non-nil, non-compatibility error because that will now fall through to the code below this block.

What if we changed this so `fsm` sets `CompatibilityError` on its own topology description and returns `f.Topology, s, nil`? Then, I don't think we need to special-case `CompatibilityError != nil` here. We should fall through to the code below, which calculates the topology difference, adds/removes servers as needed, and does `t.desc.Store(current)`, which will store the description with the updated servers and compatibility error.

Reply:
I've updated fsm.apply() to return `f.Topology, s, err` and instead of checking `CompatibilityError != nil`, I check `f.Topology.CompatibilityErr`.

Reply:
I don't think the changes in the last patch were correct. `fsm.apply` should return `f.Topology, s, nil` in the case of a compatibility error. In response, `topology.apply` doesn't need to change at all. We need to leave the `if err != nil` check as-is because that function has to be able to handle non-compatibility errors as well. If we reset it to the way it was, it will handle those errors and then apply the topology, which has the compatibility error set on it.

Reply:
done